File: packages/dashql-app/src/shell/session_relation_catalog.ts
import { CATALOG_DEFAULT_DESCRIPTOR_POOL_RANK } from '../catalog.js';
import * as core from '../core/index.js';
import type { DashQLScript } from '../core/api.js';

import type { DashQLShell } from './api.js';

const SESSION_RELATION_CATALOG_RANK = CATALOG_DEFAULT_DESCRIPTOR_POOL_RANK - 1;
const SESSION_SCHEMA = 'public';

interface SessionRelation {
    databaseName: string;
    schemaName: string;
    relationName: string;
    columns: string[];
}

interface StatementTarget {
    databaseName: string;
    schemaName: string;
    relationName: string;
}

function quoteIdentifier(identifier: string): string {
    return `"${identifier.replace(/"/g, '""')}"`;
}

function readIdentifier(identifier: string | null): string {
    return identifier?.replace(/""/g, '"') ?? '';
}

function canonicalizeTarget(target: StatementTarget): StatementTarget {
    return {
        databaseName: target.databaseName,
        schemaName: target.schemaName || SESSION_SCHEMA,
        relationName: target.relationName,
    };
}

function targetKey(target: StatementTarget): string {
    return `${target.databaseName}\0${target.schemaName}\0${target.relationName}`;
}

function readTarget(statement: core.buffers.parser.Statement): StatementTarget | null {
    const target = statement.target();
    const relationName = target?.relationName();
    if (target == null || !relationName) return null;
    return canonicalizeTarget({
        databaseName: readIdentifier(target.databaseName()),
        schemaName: readIdentifier(target.schemaName()),
        relationName: readIdentifier(relationName),
    });
}

function readCreatedRelation(
    analyzed: core.buffers.analyzer.AnalyzedScript,
    target: StatementTarget,
): SessionRelation {
    for (let i = 0; i < analyzed.tablesLength(); ++i) {
        const table = analyzed.tables(i);
        const name = table?.tableName();
        if (table == null || name == null) continue;
        const candidate = canonicalizeTarget({
            databaseName: readIdentifier(name.databaseName()),
            schemaName: readIdentifier(name.schemaName()),
            relationName: readIdentifier(name.tableName()),
        });
        if (targetKey(candidate) !== targetKey(target)) continue;

        const columns: string[] = [];
        for (let j = 0; j < table.tableColumnsLength(); ++j) {
            const columnName = table.tableColumns(j)?.columnName();
            if (columnName) columns.push(readIdentifier(columnName));
        }
        return { ...target, columns };
    }
    return { ...target, columns: [] };
}

function renderCatalog(relations: Iterable<SessionRelation>): string {
    const sorted = [...relations].sort((left, right) => targetKey(left).localeCompare(targetKey(right)));
    const statements = sorted.map(relation => {
        const name = [relation.databaseName, relation.schemaName, relation.relationName]
            .filter(part => part.length !== 0)
            .map(quoteIdentifier)
            .join('.');
        const columns = relation.columns
            .map(column => `    ${quoteIdentifier(column)} VARCHAR`)
            .join(',\n');
        return `CREATE TABLE ${name} (\n${columns}\n);`;
    });
    return ['-- Relations created during this shell session.', ...statements].join('\n\n');
}

export class ShellSessionRelationCatalog {
    private readonly parserScript: DashQLScript;
    private readonly catalogScript: DashQLScript;
    private readonly relations = new Map<string, SessionRelation>();

    constructor(private readonly shell: DashQLShell) {
        this.parserScript = shell.core.createScript(shell.catalog);
        this.catalogScript = shell.loadCatalogScript(renderCatalog([]), SESSION_RELATION_CATALOG_RANK);
    }

    /** Exposes the generated descriptor SQL for focused catalog tests. */
    getScriptText(): string {
        return this.catalogScript.toString();
    }

    applySuccessfulQuery(query: string): void {
        this.parserScript.replaceText(query);
        this.parserScript.analyze();
        const parsedBuffer = this.parserScript.getParsed();
        const analyzedBuffer = this.parserScript.getAnalyzed();
        try {
            const parsed = parsedBuffer.read();
            if (parsed.parserErrorsLength() !== 0 || parsed.scannerErrorsLength() !== 0 || parsed.statementsLength() !== 1) {
                return;
            }
            const statement = parsed.statements(0);
            if (statement == null) return;
            const target = readTarget(statement);
            if (target == null) return;

            switch (statement.statementType()) {
                case core.buffers.parser.StatementType.CREATE_TABLE:
                case core.buffers.parser.StatementType.CREATE_TABLE_AS:
                case core.buffers.parser.StatementType.CREATE_VIEW:
                case core.buffers.parser.StatementType.SELECT_INTO:
                    this.relations.set(targetKey(target), readCreatedRelation(analyzedBuffer.read(), target));
                    break;
                case core.buffers.parser.StatementType.DROP_TABLE:
                case core.buffers.parser.StatementType.DROP_VIEW:
                    this.relations.delete(targetKey(target));
                    break;
                default:
                    return;
            }
            this.reloadCatalogScript();
        } finally {
            analyzedBuffer.destroy();
            parsedBuffer.destroy();
        }
    }

    destroy(): void {
        this.shell.catalog.dropScript(this.catalogScript);
        this.parserScript.destroy();
    }

    private reloadCatalogScript(): void {
        this.catalogScript.replaceText(renderCatalog(this.relations.values()));
        this.catalogScript.analyze();
        this.shell.catalog.dropScript(this.catalogScript);
        this.shell.catalog.loadScript(this.catalogScript, SESSION_RELATION_CATALOG_RANK);
    }
}

File: packages/dashql-core/src/shell/session_relation_catalog.cc
#include "dashql/shell/session_relation_catalog.h"

#include <cstdint>
#include <optional>
#include <sstream>
#include <utility>

#include "dashql/catalog.h"
#include "dashql/script.h"

namespace dashql::shell {
namespace {

using StatementType = buffers::parser::StatementType;

constexpr uint32_t SESSION_RELATION_CATALOG_RANK = 9998;
constexpr std::string_view SESSION_SCHEMA = "public";
constexpr std::string_view CATALOG_HEADER = "-- Relations created during this shell session.";

std::string QuoteIdentifier(std::string_view identifier) {
    std::string quoted;
    quoted.reserve(identifier.size() + 2);
    quoted.push_back('"');
    for (const auto character : identifier) {
        quoted.push_back(character);
        if (character == '"') quoted.push_back('"');
    }
    quoted.push_back('"');
    return quoted;
}

struct StatementTarget {
    std::string database_name;
    std::string schema_name;
    std::string relation_name;
};

std::optional<StatementTarget> ReadTarget(const ParsedScript& parsed, size_t statement_id) {
    const auto statement = parsed.PackStatement(statement_id);
    const auto* target = statement->target.get();
    if (target == nullptr || target->relation_name.empty()) {
        return std::nullopt;
    }
    return StatementTarget{
        .database_name = target->database_name,
        .schema_name = target->schema_name.empty() ? std::string{SESSION_SCHEMA} : target->schema_name,
        .relation_name = target->relation_name,
    };
}

}  // namespace

SessionRelationCatalog::SessionRelationCatalog(Catalog& catalog)
    : catalog_{catalog},
      parser_script_{std::make_unique<Script>(catalog)},
      catalog_script_{std::make_unique<Script>(catalog)} {
    catalog_script_->ReplaceText(CATALOG_HEADER);
}

SessionRelationCatalog::~SessionRelationCatalog() = default;

void SessionRelationCatalog::ApplySuccessfulQuery(std::string_view query) {
    parser_script_->ReplaceText(query);
    parser_script_->Analyze();
    const auto& parsed = parser_script_->GetParsedScript();
    const auto& analyzed = parser_script_->GetAnalyzedScript();
    if (parsed == nullptr || analyzed == nullptr || !parsed->errors.empty() ||
        !parsed->scanned_script->errors.empty() || parsed->statements.size() != 1) {
        return;
    }

    const auto statement_type = parsed->statements.front().type;
    switch (statement_type) {
        case StatementType::CREATE_TABLE:
        case StatementType::CREATE_TABLE_AS:
        case StatementType::CREATE_VIEW:
        case StatementType::SELECT_INTO:
        case StatementType::DROP_TABLE:
        case StatementType::DROP_VIEW:
            break;
        default:
            return;
    }

    const auto target = ReadTarget(*parsed, 0);
    if (!target.has_value()) return;
    RelationKey key{target->database_name, target->schema_name, target->relation_name};

    if (statement_type == StatementType::DROP_TABLE || statement_type == StatementType::DROP_VIEW) {
        relations_.erase(key);
    } else {
        Relation relation{
            .database_name = target->database_name,
            .schema_name = target->schema_name,
            .relation_name = target->relation_name,
        };
        for (const auto& table_chunk : analyzed->GetTables().GetChunks()) {
            for (const auto& table : table_chunk) {
                if (table.table_name.database_name.get().text != relation.database_name ||
                    (table.table_name.schema_name.get().text.empty()
                         ? SESSION_SCHEMA
                         : table.table_name.schema_name.get().text) != relation.schema_name ||
                    table.table_name.table_name.get().text != relation.relation_name) {
                    continue;
                }
                relation.columns.reserve(table.table_columns.size());
                for (const auto& column : table.table_columns) {
                    relation.columns.emplace_back(column.column_name.get().text);
                }
                break;
            }
        }
        relations_.insert_or_assign(std::move(key), std::move(relation));
    }
    ReloadCatalogScript();
}

void SessionRelationCatalog::ReloadCatalogScript() {
    catalog_script_->ReplaceText(RenderCatalog());
    catalog_script_->Analyze();
    catalog_.DropScript(*catalog_script_);
    catalog_.LoadScript(*catalog_script_, SESSION_RELATION_CATALOG_RANK);
}

std::string SessionRelationCatalog::RenderCatalog() const {
    std::ostringstream output;
    output << CATALOG_HEADER;
    for (const auto& [_, relation] : relations_) {
        output << "\n\nCREATE TABLE ";
        if (!relation.database_name.empty()) output << QuoteIdentifier(relation.database_name) << '.';
        output << QuoteIdentifier(relation.schema_name) << '.' << QuoteIdentifier(relation.relation_name) << " (\n";
        for (size_t i = 0; i < relation.columns.size(); ++i) {
            if (i != 0) output << ",\n";
            output << "    " << QuoteIdentifier(relation.columns[i]) << " VARCHAR";
        }
        output << "\n);";
    }
    return output.str();
}

}  // namespace dashql::shell
